text/353-cfn-registry-constructs.md

```diff
@@ -0,0 +1,257 @@
+use them via strongly-typed classes. IDEs will show type information and inline help derived from the
+extension schema.
+
+```ts
+import { CfnCluster } from '@cdk-cloudformation-types/awsqs-eks-cluster';
+
+new CfnCluster(this, 'MyEksCluster', {
+  name: 'my-new-cluster',
+  tags: [ { key: 'foo', value: 'bar' } ],
+});
+```
+
+## Internal FAQ
+
+### Why are we doing this?
+
+We are doing this because CDK users would like to be able to use CloudFormation extensions
+(modules, resources) as first-class citizens in CDK code. Extensions and native resource types
+should look and feel the same way when defined in CDK apps.
+
+Additionally, we would like to surface CloudFormation extensions in the upcoming Construct Hub
+and to that end, if we simply publish a construct library that includes constructs for all the public,
+extensions, this library will naturally be discoverable through the Hub.
+
+### Why should we _not_ do this?
+
+We have a longer term plan ([RFC#77](https://github.com/aws/aws-cdk-rfcs/issues/77)) to add
+support for a CLI command called `cdk import` which will allow users to generate L1s from any
+registry type just-in-type and add them to their project (this is similar to CDK8s and CDKtf).
+
+### What is the technical solution (design) of this feature?
+
+The idea is to do something similar to what we do with the CloudFormation L1s
+in the AWS CDK and simply publish JSII construct libraries which includes
+statically-generated L1s for all the public registry extensions.
+
+#### Package per type
+
+CloudFormation extensions are semantically versioned. This means that new versions of an extension
+may include breaking changes. We also want to allow users to pick up a specific version of an extension.
+In light of these constraints, we determined that the best path forward is to publish each extension (type)
+as a separate module, with a version that corresponds to the extension version in the CloudFormation
+registry. This way, we will have 1:1 alignment and users will have the most freedom.
+
+There are currently 44 public types (resources & modules), but we expect this
+list to substantially grow. If this dramatically scales (to thousands of
+packages) and we hit any limitations of package managers (e.g. account/scope
+limits), we can shard the solution (multiple accounts, multiple scopes, etc).
+
+#### Naming Scheme
+
+> *This is still not finalized. From a CDK user standpoint, I am not sure if they should really
+> care if these are CloudFormation registry extensions or just normal CDK libraries. Consider a user
+> that goes to Construct Hub and searches for "mongodb". They find this library, import it and use it.
+> As far as they are concerned, they just used a construct library. The fact that this is an L1 type
+> and not some L2/L3 is an implementation details, isn't it. So I am not sure that the semantics which
```

Comment:
While in principle I agree, there are practical differences.

People have gotten used to certain affordances in L2s that these constructs won't have. Typed arguments instead of ARNs (`bucket: IBucket` vs `bucketArn: string`), integration classes, enums, a README with a thorough explanation and examples, etc.

Since these classes will be generated at the L1 level, the `Cfn` prefix will indicate that all these expected affordances are absent.

Reply:
On the other hand, the `Cfn` prefix might be taken to mean there *should* be a better, preferred version of the construct... which will never arrive.

Reply:
I am coming down on the side of NO `Cfn` prefix.

Reply:
Why not use The Cfn prefix?

Reply:
@rix0rrr and I discussed this and decided we will keep the `Cfn` prefix in order to allow future publishing of 1st party L1s as part of this project. In which case we will likely have a library such as `@cdk-cloudformation/aws-lambda.CfnFunction`.